src/ai-jsx/util.ts

```diff
@@ -0,0 +1,372 @@
+  } else if (span.name === makeComponentName(OpenAIChatModel)) {
+    // There is a nested completion not wrapped in a placeholder
+    const name = parseTrackerIdFromProps(span.props) || 'completion';
+    return `{{ ${name} }}`;
+  } else {
+    return span.children.map(makeTemplateString).filter(Boolean).join('');
+  }
+}
+
+function makeMessageString(span: ABSpan | string): string {
+  if (typeof span === 'string') {
+    return span;
+  } else {
+    return span.children.map(makeMessageString).filter(Boolean).join('');
+  }
+}
+
+function countMessageTokens(span: ABSpan | string): number {
+  let tokens = 0;
+  if (typeof span === 'string') {
+    tokens++;
+  } else {
+    span.children.forEach((child) => {
+      tokens += countMessageTokens(child);
+    });
+  }
+  return tokens;
+}
+
+interface AIMessage {
+  role: 'system' | 'user' | 'assistant';
+  content: string;
+  tokens?: number;
+}
+
+interface MemoizedMessage {
+  memoizedId: string | undefined;
+  message: AIMessage;
+}
+
+function makeMemoizedMessagesArray(
+  span: ABSpan,
+  components: AnyComponent[],
+): MemoizedMessage[] {
+  const componentNames = components.map(makeComponentName);
+  const messages: MemoizedMessage[] = [];
+
+  let content: string | undefined = undefined;
+  let role: AIMessage['role'] | undefined = undefined;
+  let tokens: number | undefined = undefined;
+
+  if (componentNames.includes(span.name)) {
+    content = makeMessageString(span);
+
+    switch (span.name) {
+      case makeComponentName(SystemMessage):
+        role = 'system';
+        break;
+      case makeComponentName(UserMessage):
+        role = 'user';
+        break;
+      case makeComponentName(AssistantMessage):
+        role = 'assistant';
+        tokens = countMessageTokens(span);
+        break;
+    }
+  }
+
+  if (content && role) {
+    messages.push({
+      memoizedId: span.memoizedId,
+      message: { content, role, tokens },
+    });
+  }
+
+  for (const child of span.children) {
+    if (typeof child === 'string') {
+      continue;
+    }
+
+    // Stop if we encounter another chat component
+    if (child.name === makeComponentName(OpenAIChatModel)) {
+      continue;
+    }
+
+    messages.push(...makeMemoizedMessagesArray(child, components));
+  }
+
+  return messages;
+}
+
+/**
+ * Keep the last message from each memoizedId.
+ */
+function makeMessagesArrayFrom(
+  span: ABSpan,
+  components: AnyComponent[],
+): AIMessage[] {
+  const memoized = makeMemoizedMessagesArray(span, components);
+
+  const seenMemoizedIds = new Set<string>();
+  const dedupedMessages: AIMessage[] = [];
+
+  for (const memo of memoized.reverse()) {
+    if (memo.memoizedId === undefined) {
+      dedupedMessages.push(memo.message);
+      continue;
+    }
+
+    if (seenMemoizedIds.has(memo.memoizedId)) {
+      continue;
+    }
+
+    seenMemoizedIds.add(memo.memoizedId);
+    dedupedMessages.push(memo.message);
+  }
+
+  return dedupedMessages.reverse();
```

Comment:
I think you can use unshift here instead of push so you don't have to reverse the array.